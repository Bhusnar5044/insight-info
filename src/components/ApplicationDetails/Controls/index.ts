export { Controls as Default } from './Controls';
