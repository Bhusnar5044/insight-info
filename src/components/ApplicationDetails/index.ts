export { ApplicationDetails as default } from './ApplicationDetails';
