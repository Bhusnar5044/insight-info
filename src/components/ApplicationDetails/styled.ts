import { Box } from '@components/common';
import { theme } from '@theme';
import { styled } from '@utils';

export const CardContainer = styled(Box)`
    width: 90%;
    margin: 2rem auto;
`;
export const Card = styled(Box)`
    min-width: 22rem;
    padding: 2rem;
    text-align: left;
    color: ${theme.colors.other.purpleTypo};
    background-color: ${theme.colors.other.lightPurple};
    border: 0.1rem solid ${theme.colors.other.lightPurple};
    border-radius: 0.5rem;
    &:hover {
        border: 0.1rem solid ${theme.colors.other.lightPurple};
        box-shadow: rgba(100, 100, 111, 0.2) 0px 7px 29px 0px;
    }
`;
