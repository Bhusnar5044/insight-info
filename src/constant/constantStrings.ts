export const emptyStateMessages = {
    application: 'No applications available.',
    resources: 'No Resources available.',
    noResultFound: 'No result found',
};
