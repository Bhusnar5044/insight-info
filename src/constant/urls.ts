const baseURL = 'https://engineering-task.elancoapps.com/api';

export const urls = {
    login: 'https://api.escuelajs.co/api/v1/auth/login',
    profile: 'https://api.escuelajs.co/api/v1/auth/profile',
    raw: `${baseURL}/raw`,
    applications: `${baseURL}/applications`,
    resources: `${baseURL}/resources`,
};
