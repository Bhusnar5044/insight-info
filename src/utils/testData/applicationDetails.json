[
    {
        "ConsumedQuantity": "2",
        "Cost": "21.5424",
        "Date": "29/11/2020",
        "InstanceId": "dummy",
        "MeterCategory": "dummy category",
        "ResourceGroup": "dummy group",
        "ResourceLocation": "dummy location",
        "Tags": {
            "app-name": "dummy app name",
            "environment": "test",
            "business-unit": "dummy business unit"
        },
        "UnitOfMeasure": "1 Hour",
        "Location": "dummy location",
        "ServiceName": "dummy Service name"
    },
    {
        "ConsumedQuantity": "3",
        "Cost": "21.5424",
        "Date": "30/11/2020",
        "InstanceId": "dummy",
        "MeterCategory": "dummy category",
        "ResourceGroup": "dummy group",
        "ResourceLocation": "dummy location",
        "Tags": {
            "app-name": "dummy app name",
            "environment": "production",
            "business-unit": "dummy business unit"
        },
        "UnitOfMeasure": "1 Hour",
        "Location": "dummy location",
        "ServiceName": "dummy Service name"
    }
]
