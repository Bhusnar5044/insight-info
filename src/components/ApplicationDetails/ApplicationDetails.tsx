import { Loader } from '@components/Loader/Loader';
import { Box, Text } from '@components/common';
import { urls } from '@constant/urls';
import { fetch } from '@utils';
import { ChangeEvent, FC, memo, useCallback, useEffect, useState } from 'react';
import { useParams } from 'react-router-dom';
import { Controls } from './Controls/Controls';
import { DateRange } from './Controls/types';
import { Card, CardContainer } from './styled';
import { IApplicationDetails } from './types';

export const ApplicationDetails: FC = memo(() => {
    const { category = 'applications', groupName } = useParams();
    const [isLoading, setIsLoading] = useState(false);
    const [applicationDetails, setApplicationDetails] = useState<IApplicationDetails[]>([]);
    const [filteredApplicationDetails, setFilteredApplicationDetails] =
        useState<IApplicationDetails[]>(applicationDetails);
    const [sort, setSort] = useState('');
    const [dateRange, setDateRange] = useState<DateRange>([new Date(), new Date()]);

    const getApplicationDetails = useCallback(async () => {
        setIsLoading(true);
        let url = '';
        if (category === 'applications') url = `${urls.applications}/${groupName}`;
        else if (category === 'resources') url = `${urls.resources}/${groupName}`;

        const { response, error } = await fetch({ url });
        const data = response?.data;
        if (response?.data) {
            for (const item of data) {
                const dateArr = item.Date.split('/');
                const year = parseFloat(dateArr[2]);
                const month = parseFloat(dateArr[1]) - 1;
                const day = parseFloat(dateArr[0]);
                const newDate = new Date(year, month, day);
                // Update the object
                item.Date = newDate;
            }
            setApplicationDetails(data);
            setFilteredApplicationDetails(data);
            if (data.length) {
                const sortedData = [...data];
                sortedData.sort((a, b) => a.Date.getTime() - b.Date.getTime());
                setDateRange([sortedData[0].Date, sortedData[data.length - 1].Date]);
            }
        }
        if (error) console.log(error);

        setIsLoading(false);
    }, [category, groupName]);

    const onSearch = (value: string) => {
        const search = value.toLowerCase();
        setFilteredApplicationDetails(
            applicationDetails.filter((item: IApplicationDetails) => {
                if (
                    item.Cost.includes(search) ||
                    item.UnitOfMeasure.includes(search) ||
                    item.ServiceName.toLowerCase().includes(search) ||
                    item.Location.toLowerCase().includes(search) ||
                    item.Tags.environment.toLowerCase().includes(search)
                ) {
                    console.log({ search, item });
                    return true;
                }
                return false;
            }),
        );
    };

    const onSortChange = useCallback(
        (e: ChangeEvent<HTMLSelectElement>) => {
            const value = e.target.value;
            console.log(value);
            setSort(value);
            const sortedList = [...applicationDetails];
            if (value === 'cost-DSC') sortedList.sort((a, b) => b.Cost.localeCompare(a.Cost));
            else if (value === 'cost-ASC') sortedList.sort((a, b) => b.Cost.localeCompare(a.Cost)).reverse();
            else if (value === 'date-DSC') {
                sortedList.sort((a, b) => b.Date.getTime() - a.Date.getTime());
            } else if (value === 'date-ASC') sortedList.sort((a, b) => a.Date.getTime() - b.Date.getTime());
            setFilteredApplicationDetails(sortedList);
        },
        [applicationDetails],
    );

    const handleDateChange = useCallback(
        (value: DateRange) => {
            setDateRange(value);
            const [startDate, endDate] = value as Date[];
            console.log({ startDate, endDate });
            const sortedList = [...applicationDetails];
            sortedList
                .sort((a, b) => a.Date.getTime() - b.Date.getTime())
                .filter(
                    (item) => item.Date.getTime() >= startDate.getTime() && item.Date.getTime() <= endDate.getTime(),
                );
            setFilteredApplicationDetails(sortedList);
        },
        [applicationDetails],
    );

    useEffect(() => {
        getApplicationDetails();
    }, [getApplicationDetails]);
    return (
        <CardContainer flexDirection="column" gap="2rem">
            <Text textVariant="h3" textColor="#fff" textAlign="center">
                {groupName}
            </Text>
            <Controls
                onSearch={onSearch}
                onSortChange={onSortChange}
                sort={sort}
                handleDateChange={handleDateChange}
                dateRange={dateRange}
            />
            <CardContainer gap="2rem" flexWrap="wrap">
                {isLoading && <Loader />}
                {!isLoading &&
                    filteredApplicationDetails.map((item) => (
                        <Card
                            key={item.Date.toLocaleDateString('en-GB')}
                            flexDirection="column"
                            gap="1rem"
                            alignItems="start">
                            <Text textVariant="body3">
                                <strong>Date:</strong> {item.Date.toLocaleDateString('en-GB')}
                            </Text>
                            <Text textVariant="body3">
                                <strong>Cost:</strong> {item.Cost}
                            </Text>
                            <Text textVariant="body3">
                                <strong>Meter category:</strong> {item.MeterCategory}
                            </Text>
                            <Text textVariant="body3">
                                <strong>Location:</strong> {item.Location}
                            </Text>
                            <Text textVariant="body3">
                                <strong>Unit of measure:</strong> {item.UnitOfMeasure}
                            </Text>
                            <Text textVariant="body3" textWeight="Strong">
                                Tags:
                            </Text>
                            <Box flexDirection="column" margin="0 0 0 1rem" alignItems="start" gap="1rem">
                                <Text textVariant="body3">
                                    <strong>app-name:</strong> {item.Tags['app-name']}
                                </Text>
                                <Text textVariant="body3">
                                    <strong>environment:</strong> {item.Tags['environment']}
                                </Text>
                                <Text textVariant="body3">
                                    <strong>business-unit:</strong> {item.Tags['business-unit']}
                                </Text>
                            </Box>
                        </Card>
                    ))}
            </CardContainer>
        </CardContainer>
    );
});
