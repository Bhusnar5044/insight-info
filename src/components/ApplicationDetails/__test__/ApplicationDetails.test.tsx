import { emptyStateMessages } from '@constant/constantStrings';
import { urls } from '@constant/urls';
import { fireEvent, mockAxios, renderWithRouter, screen, waitFor } from '@test-utils';
import * as mockData from '@utils/testData/applicationDetails.json';
import { ApplicationDetails } from '../ApplicationDetails';

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'), // use actual for all non-hook parts
    useParams: () => ({
        category: 'applications',
        groupName: 'dummy',
    }),
}));

const renderer = () => renderWithRouter(<ApplicationDetails />);
describe('Dashboard', () => {
    beforeEach(() => {
        jest.resetModules();
        jest.resetAllMocks();
        mockAxios.resetHistory();
    });

    it('should render properly', () => {
        const { container } = renderer();
        expect(container).toMatchSnapshot();
        expect(screen.getByText(/dummy/i)).toBeInTheDocument();
    });

    it('should show loader', () => {
        renderer();
        expect(screen.getByTitle('loader')).toBeInTheDocument();
    });

    it('should show no result available message on failure or in case response is empty', async () => {
        mockAxios.onGet(`${urls.applications}/dummy`).reply(200, []);
        renderer();
        expect(screen.getByTitle('loader')).toBeInTheDocument();
        await waitFor(() => {
            expect(screen.queryByTitle('loader')).not.toBeInTheDocument();
            expect(screen.getByText(emptyStateMessages.noResultFound)).toBeInTheDocument();
        });
    });

    it('should fetch all the applications', async () => {
        mockAxios.onGet(`${urls.applications}/dummy`).reply(200, mockData);
        renderer();
        expect(screen.getByTitle('loader')).toBeInTheDocument();
        await waitFor(() => {
            expect(screen.queryByTitle('loader')).not.toBeInTheDocument();
            expect(screen.getByText(/production/i)).toBeInTheDocument();
        });
    });

    it('should show the result based on search', async () => {
        mockAxios.onGet(`${urls.applications}/dummy`).reply(200, mockData);
        renderer();
        expect(screen.getByTitle('loader')).toBeInTheDocument();
        await waitFor(() => {
            expect(screen.queryByTitle('loader')).not.toBeInTheDocument();
            expect(screen.getByText(/test/i)).toBeInTheDocument();
            const searchBox = screen.getByTitle('searchBox');
            fireEvent.change(searchBox, { target: { value: 'production' } });
            fireEvent.keyDown(searchBox, { key: 'Enter' });
            screen.debug();
            expect(screen.queryByTitle(/test/i)).not.toBeInTheDocument();
            expect(screen.getByText(/production/i)).toBeInTheDocument();
        });
    });
});
