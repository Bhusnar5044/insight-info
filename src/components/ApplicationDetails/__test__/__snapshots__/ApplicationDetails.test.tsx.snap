// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard should render properly 1`] = `
.c12 {
  width: 10rem;
  height: auto;
  margin: auto;
  animation: gOaCMp infinite 20s linear;
}

.c0 {
  display: flex;
  flex-direction: column;
  gap: 2rem;
}

.c3 {
  display: flex;
  gap: 1rem;
}

.c8 {
  display: flex;
  gap: 0.5rem;
}

.c11 {
  display: flex;
  justify-content: center;
  flex-wrap: wrap;
  gap: 2rem;
}

.c7 {
  -webkit-appearance: none;
  appearance: none;
  width: 100%;
  border: 0;
  font-family: inherit;
  font-size: 16px;
  font-weight: 400;
  background: #ffff;
  box-shadow: inset 0 -1px 0 rgba(#212529, 0.3);
  color: #212529;
  transition: all 0.15s ease;
  outline: none;
}

.c5 {
  position: relative;
  border-radius: 3px;
  padding: 0.5rem 1rem;
  display: flex;
  align-items: center;
  overflow: hidden;
  width: 23rem;
  background-color: #ffff;
}

.c5 .focus-bg {
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background: rgba(#6c757d, 0.05);
  z-index: -1;
  transform: scaleX(0);
  transform-origin: left;
}

.c5 .c6 {
  width: 20rem;
}

.c5 img {
  cursor: pointer;
  width: 1.2rem;
  height: 1.2rem;
}

.c9 {
  background-color: white;
  border-radius: 0.3rem;
  display: inline-block;
  font: inherit;
  line-height: 1.5rem;
  padding: 0.5rem 3.5rem 0.5rem 1rem;
  margin: 0;
  outline: unset;
  -webkit-box-sizing: border-box;
  -moz-box-sizing: border-box;
  box-sizing: border-box;
  -webkit-appearance: none;
  -moz-appearance: none;
  background-image: linear-gradient(45deg, transparent 50%, gray 50%),linear-gradient(135deg, gray 50%, transparent 50%),radial-gradient(#ddd 70%, transparent 72%);
  background-position: calc(100% - 20px) calc(1em + 2px),calc(100% - 15px) calc(1em + 2px),calc(100% - 0.5em) 0.5em;
  background-size: 5px 5px,5px 5px,1.5em 1.5em;
  background-repeat: no-repeat;
}

.c9 :focus {
  background-image: linear-gradient(45deg, white 50%, transparent 50%),linear-gradient(135deg, transparent 50%, white 50%),radial-gradient(gray 70%, transparent 72%);
  background-position: calc(100% - 15px) 1em,calc(100% - 20px) 1em,calc(100% - 0.5em) 0.5em;
  background-size: 5px 5px,5px 5px,1.5em 1.5em;
  background-repeat: no-repeat;
}

.c10 {
  padding: 0.5rem;
}

.c2 {
  margin: 0;
  color: #fff;
  font-size: 2.4rem;
  font-weight: 400;
  letter-spacing: -0.04rem;
  line-height: 3.2rem;
  text-align: center;
}

.c4 {
  width: 90%;
  align-items: baseline;
  justify-content: center;
}

.c4 .react-daterange-picker__wrapper {
  background-color: #fff;
  padding: 0.4rem 1rem;
  border-radius: 0.3rem;
}

.c1 {
  width: 90%;
  margin: 2rem auto;
}

<div>
  <div
    class="c0 c1"
    flexdirection="column"
    gap="2rem"
  >
    <h3
      class="c2"
      textalign="center"
      textcolor="#fff"
      textvariant="h3"
    >
      dummy
    </h3>
    <div
      class="c3 c4"
      gap="1rem"
    >
      <div
        class="c5"
        for="searchInput"
      >
        <input
          class="c6 c7"
          id="searchInput"
          placeholder="Search"
          title="searchBox"
          type="text"
          value=""
        />
        <div
          class="c8"
          gap="0.5rem"
        >
          <img
            alt="search-icon"
          />
        </div>
      </div>
      <div
        class="c3"
        gap="1rem"
      >
        <select
          class="c9"
          title="single-select"
        >
          <option
            class="c10"
            value=""
          >
            Select sort
          </option>
          <option
            class="c10"
            value="cost-DSC"
          >
            Sort by Cost (H-L)
          </option>
          <option
            class="c10"
            value="cost-ASC"
          >
            Sort by Cost (L-H)
          </option>
          <option
            class="c10"
            value="date-DSC"
          >
            Sort by Date (DSC)
          </option>
          <option
            class="c10"
            value="date-ASC"
          >
            Sort by Date (ASC)
          </option>
        </select>
      </div>
      <div
        class="react-daterange-picker react-daterange-picker--closed react-daterange-picker--enabled"
      >
        <div
          class="react-daterange-picker__wrapper"
        >
          <div
            class="react-daterange-picker__inputGroup"
          >
            <input
              hidden=""
              max="275760-09-13"
              min="0001-01-01"
              name="daterange_from"
              style="visibility: hidden; position: absolute; z-index: -999;"
              type="date"
              value="2023-06-17"
            />
            <input
              autocomplete="off"
              class="react-daterange-picker__inputGroup__input react-daterange-picker__inputGroup__month"
              data-input="true"
              inputmode="numeric"
              max="12"
              min="1"
              name="month"
              placeholder="--"
              type="number"
              value="6"
            />
            <span
              class="react-date-picker__inputGroup__divider"
            >
              /
            </span>
            <input
              autocomplete="off"
              class="react-daterange-picker__inputGroup__input react-daterange-picker__inputGroup__day"
              data-input="true"
              inputmode="numeric"
              max="30"
              min="1"
              name="day"
              placeholder="--"
              type="number"
              value="17"
            />
            <span
              class="react-date-picker__inputGroup__divider"
            >
              /
            </span>
            <input
              autocomplete="off"
              class="react-daterange-picker__inputGroup__input react-daterange-picker__inputGroup__year"
              data-input="true"
              inputmode="numeric"
              max="275760"
              min="1"
              name="year"
              placeholder="----"
              step="1"
              type="number"
              value="2023"
            />
          </div>
          <span
            class="react-daterange-picker__range-divider"
          >
            â€“
          </span>
          <div
            class="react-daterange-picker__inputGroup"
          >
            <input
              hidden=""
              max="275760-09-13"
              min="0001-01-01"
              name="daterange_to"
              style="visibility: hidden; position: absolute; z-index: -999;"
              type="date"
              value="2023-06-17"
            />
            <input
              autocomplete="off"
              class="react-daterange-picker__inputGroup__input react-daterange-picker__inputGroup__month"
              data-input="true"
              inputmode="numeric"
              max="12"
              min="1"
              name="month"
              placeholder="--"
              type="number"
              value="6"
            />
            <span
              class="react-date-picker__inputGroup__divider"
            >
              /
            </span>
            <input
              autocomplete="off"
              class="react-daterange-picker__inputGroup__input react-daterange-picker__inputGroup__day"
              data-input="true"
              inputmode="numeric"
              max="30"
              min="1"
              name="day"
              placeholder="--"
              type="number"
              value="17"
            />
            <span
              class="react-date-picker__inputGroup__divider"
            >
              /
            </span>
            <input
              autocomplete="off"
              class="react-daterange-picker__inputGroup__input react-daterange-picker__inputGroup__year"
              data-input="true"
              inputmode="numeric"
              max="275760"
              min="1"
              name="year"
              placeholder="----"
              step="1"
              type="number"
              value="2023"
            />
          </div>
          <button
            class="react-daterange-picker__clear-button react-daterange-picker__button"
            type="button"
          >
            <svg
              class="react-daterange-picker__clear-button__icon react-daterange-picker__button__icon"
              height="19"
              stroke="black"
              stroke-width="2"
              viewBox="0 0 19 19"
              width="19"
              xmlns="http://www.w3.org/2000/svg"
            >
              <line
                x1="4"
                x2="15"
                y1="4"
                y2="15"
              />
              <line
                x1="15"
                x2="4"
                y1="4"
                y2="15"
              />
            </svg>
          </button>
          <button
            class="react-daterange-picker__calendar-button react-daterange-picker__button"
            type="button"
          >
            <svg
              class="react-daterange-picker__calendar-button__icon react-daterange-picker__button__icon"
              height="19"
              stroke="black"
              stroke-width="2"
              viewBox="0 0 19 19"
              width="19"
              xmlns="http://www.w3.org/2000/svg"
            >
              <rect
                fill="none"
                height="15"
                width="15"
                x="2"
                y="2"
              />
              <line
                x1="6"
                x2="6"
                y1="0"
                y2="4"
              />
              <line
                x1="13"
                x2="13"
                y1="0"
                y2="4"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
    <div
      class="c11 c1"
      flexwrap="wrap"
      gap="2rem"
      justifycontent="center"
    >
      <img
        class="c12"
        title="loader"
      />
    </div>
  </div>
</div>
`;
